src/java/org/apache/cassandra/index/sai/StorageAttachedIndexBuilder.java

```diff
@@ -0,0 +1,351 @@
+            // build the per-index components
+            boolean perIndexComponentsOnly = perSSTableFileLock == null;
+            // remove existing per column index files instead of overwriting
+            IndexDescriptor indexDescriptor = IndexDescriptor.create(sstable);
+            indexes.forEach(index -> indexDescriptor.deleteColumnIndex(index.getIndexContext()));
+
+            indexWriter = new StorageAttachedIndexWriter(indexDescriptor, indexes, txn, perIndexComponentsOnly);
+
+            long previousKeyPosition = 0;
+            indexWriter.begin();
+
+            try (KeyIterator keys = new KeyIterator(sstable.descriptor, metadata))
+            {
+                while (keys.hasNext())
+                {
+                    if (isStopRequested())
+                    {
+                        logger.debug(indexDescriptor.logMessage("Index build has been stopped"));
+                        throw new CompactionInterruptedException(getCompactionInfo());
+                    }
+
+                    final DecoratedKey key = keys.next();
+                    final long keyPosition = keys.getKeyPosition();
+
+                    indexWriter.startPartition(key, keyPosition);
+
+                    RowIndexEntry<?> indexEntry = sstable.getPosition(key, SSTableReader.Operator.EQ);
+                    dataFile.seek(indexEntry.position);
+                    ByteBufferUtil.readWithShortLength(dataFile); // key
+
+                    try (SSTableIdentityIterator partition = SSTableIdentityIterator.create(sstable, dataFile, key))
+                    {
+                        // if the row has statics attached, it has to be indexed separately
+                        if (metadata.hasStaticColumns())
+                            indexWriter.nextUnfilteredCluster(partition.staticRow());
+
+                        while (partition.hasNext())
+                            indexWriter.nextUnfilteredCluster(partition.next());
+                    }
+
+                    bytesProcessed += keyPosition - previousKeyPosition;
+                    previousKeyPosition = keyPosition;
+                }
+
+                completeSSTable(indexWriter, sstable, indexes, perSSTableFileLock);
+            }
+
+            return false;
+        }
+        catch (Throwable t)
+        {
+            if (indexWriter != null)
+            {
+                indexWriter.abort(t, true);
+            }
+
+            if (t instanceof InterruptedException)
+            {
+                // TODO: Is there anything that makes more sense than just restoring the interrupt?
+                logger.warn(logMessage("Interrupted while building indexes {} on SSTable {}"), indexes, sstable.descriptor);
+                Thread.currentThread().interrupt();
+                return true;
+            }
+            else if (t instanceof CompactionInterruptedException)
+            {
+                //TODO Shouldn't do this if the stop was interrupted by a truncate
```

Comment:
Shouldn't do what? Throw if it's an initial build?

Reply:
Again, I will clarify this in the tests. I'm assuming it means that we shouldn't get this exception if the stop was interrupted by a truncate but it isn't very clear.